library/alloc/src/vec/spec_from_iter_nested.rs

```diff
@@ -23,9 +26,22 @@ where
         let mut vector = match iterator.next() {
             None => return Vec::new(),
             Some(element) => {
-                let (lower, _) = iterator.size_hint();
-                let mut vector = Vec::with_capacity(lower.saturating_add(1));
+                // Do -1 because we would add 1 before `with_capacity` call.
+                let min_non_zero_cap_sub_1 = RawVec::<T>::MIN_NON_ZERO_CAP.saturating_sub(1);
+                let guess_rest_len = match iterator.size_hint() {
+                    // Trivial case.
+                    (lower, Some(upper)) if lower == upper => lower,
+                    // Make default allocation as for `Vec::push` if we lack information.
+                    (0, Some(upper)) => cmp::min(min_non_zero_cap_sub_1, upper),
+                    (lower, None) => cmp::max(min_non_zero_cap_sub_1, lower),
+                    // Try to guess capacity if we get at least 1 more item than `lower`.
+                    // This would cause reallocation to at least lower*2
+                    // so let's allocate it beforehand.
+                    (lower, Some(upper)) => cmp::min(lower * 2, upper),
+                };
+                let mut vector = Vec::with_capacity(guess_rest_len.saturating_add(1));
```

Comment:
I think the idea of amortizing tiny with_capacity's to at least MIN_NON_ZERO_CAP makes sense, but I would suggest trying `with_capacity(max(MIN_NON_ZERO_CAP, lower.wrapping_add(1))` which should be roughly equivalent to the previous code, just bumping the smallest allocations up to 4-8 elements if we'd expect to do so on reserve.

As a passing thought, it may be worth considering a `Vec::with_capacity_hint` API, which is the Vec::reserve equivalent of Vec::with_capacity's reserve_exact.

Reply:
I changed code as you suggested. May you run perf again, please?